native-gradle-plugin/src/functionalTest/groovy/org/graalvm/buildtools/gradle/LayeredApplicationFunctionalTest.groovy

```diff
@@ -0,0 +1,155 @@
+import org.graalvm.buildtools.gradle.fixtures.GraalVMSupport
+import org.graalvm.buildtools.utils.NativeImageUtils
+import spock.lang.Requires
+import spock.util.concurrent.PollingConditions
+
+import java.net.http.HttpClient
+import java.net.http.HttpRequest
+import java.net.http.HttpResponse
+import java.nio.charset.StandardCharsets
+
+@Requires(
+        { NativeImageUtils.getMajorJDKVersion(GraalVMSupport.getGraalVMHomeVersionString()) >= 25 }
+)
+class LayeredApplicationFunctionalTest extends AbstractFunctionalTest {
+    def "can build a native image using layers"() {
+        def nativeApp = getExecutableFile("build/native/nativeCompile/layered-java-application")
+
+        given:
+        withSample("layered-java-application")
+
+        when:
+        run 'nativeLibdependenciesCompile'
+
+        then:
+        tasks {
+            succeeded ':nativeLibdependenciesCompile'
+        }
+        outputContains "'-H:LayerCreate' (origin(s): command line)"
+
+        when:
+        run 'nativeRun', '-Pmessage="Hello, layered images!"'
+
+        then:
+        tasks {
+            upToDate ':nativeLibdependenciesCompile'
+            succeeded ':nativeCompile'
+        }
+        nativeApp.exists()
+
+        and:
+        outputContains "- '-H:LayerUse' (origin(s): command line)"
+        outputContains "Hello, layered images!"
+
+        when: "Updating the application without changing the dependencies"
+        file("src/main/java/org/graalvm/demo/Application.java").text = """
+package org.graalvm.demo;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class Application {
+    private static final Logger LOGGER = LoggerFactory.getLogger(Application.class);
+
+    public static void main(String[] args) {
+        LOGGER.info("App started with args {}", String.join(", ", args));
+    }
+
+}
+
+"""
+        run 'nativeRun', '-Pmessage="Hello, layered images!"'
+
+        then:
+        tasks {
+            // Base layer is not rebuilt
+            upToDate ':nativeLibdependenciesCompile'
+            // Application layer is recompiled
+            succeeded ':nativeCompile'
+        }
+
+        outputContains "- '-H:LayerUse' (origin(s): command line)"
+        outputContains "Hello, layered images!"
+    }
+
+    def "can build a layered Micronaut application"() {
+        given:
+        withSample("layered-mn-application")
+
+        when:
+        run 'nativeCompile'
+
+        then:
+        tasks {
+            succeeded ':nativeLibdependenciesCompile', ':nativeCompile'
+        }
+
+        when:
+        def builder = new ProcessBuilder()
+            .directory(testDirectory.toFile())
+            .inheritIO()
+            .command("build/native/nativeCompile/layered-mn-app${IS_WINDOWS?".exe":""}")
+        def env = builder.environment()
+        env["LD_LIBRARY_PATH"] = testDirectory.resolve("build/native/nativeLibdependenciesCompile").toString()
```

Comment:
Instead of setting `LD_LIBRARY_PATH` you can set `-H:LinkerRPath={layer_path}` when building the app layer to point to the directory containing the shared object(s) of underlying layer(s). The value gets passed to the linker as the `-rpath`. The path is relative to the directory where the executable resides. For example `-H:LinkerRPath=$ORIGIN/lib` points to the `lib` directory next to the executable.

Reply:
I had tried this without success. That said I'm not sure it's a good idea: this is only used for running the app, it is not required when building. So if we change this to use `-H:LinkerRPath`, wouldn't that mean that the binary wouldn't be portable anymore if the base layer paths change?

Reply:
That's correct, if you set `-rpath` then that gets hardcoded in the binary, but it's common for apps to depend on dirs relative to the main binary. I remember you had issues with `LD_LIBRARY_PATH` before, that's why I mentioned it. Also, you need to always be sure to not overwrite `LD_LIBRARY_PATH` in case other apps depend on it.